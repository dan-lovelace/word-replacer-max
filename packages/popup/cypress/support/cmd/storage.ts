Cypress.Commands.add("getBrowser", () => {
  cy.window().then((win) => {
    cy.wrap(win.CYPRESS_BROWSER).as("browser");
  });
});
