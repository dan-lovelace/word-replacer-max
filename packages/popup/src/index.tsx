import { render } from "preact";
import { LocationProvider, Router, Route } from "preact-iso";

import { POPUP_ROUTES } from "@worm/shared";

import Layout from "./containers/Layout";
import ToastContainer from "./components/alert/ToastContainer";
import HomePage from "./pages/Home";
import NotFoundPage from "./pages/NotFound";
import { ConfigProvider } from "./store/Config";

import "./style/index.scss";
import "material-icons/iconfont/sharp.scss";

export function App() {
  return (
    <>
      <ConfigProvider>
        <LocationProvider>
          <Layout>
            <Router>
              <Route path={POPUP_ROUTES.HOME} component={HomePage} />
              <Route default component={NotFoundPage} />
            </Router>
          </Layout>
        </LocationProvider>
      </ConfigProvider>
      <ToastContainer />
    </>
  );
}

render(<App />, document.getElementById("app") as HTMLElement);
