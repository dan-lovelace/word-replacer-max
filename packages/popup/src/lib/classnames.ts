export const COPY_CONTAINER_COL_CLASS = "col col-sm-8 col-xxl-5";
