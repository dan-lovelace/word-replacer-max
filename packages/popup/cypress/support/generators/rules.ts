import { merge } from "ts-deepmerge";

import { MatcherInSync } from "@worm/types/src/rules";

export const TEST_MATCHER_ID_1 = "matcher__1234";
export const TEST_MATCHER_ID_2 = "matcher__5678";

export const testRules: MatcherInSync = {
  [TEST_MATCHER_ID_1]: {
    active: true,
    identifier: "1234",
    queries: ["my jaw dropped", "I was shocked"],
    queryPatterns: [],
    replacement: "I was surprised",
  },
  [TEST_MATCHER_ID_2]: {
    active: true,
    identifier: "5678",
    queries: ["This."],
    queryPatterns: ["case", "wholeWord"],
    replacement: " ",
  },
};

export const generateMatchers = (overrides: MatcherInSync = {}) =>
  merge(testRules, overrides);
