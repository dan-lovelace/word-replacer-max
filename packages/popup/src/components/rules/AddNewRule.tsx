import { v4 as uuidv4 } from "uuid";

import { storageSetByKeys } from "@worm/shared/src/storage";

import { PreactChildren } from "../../lib/types";
import { useConfig } from "../../store/Config";

import { useToast } from "../alert/useToast";

type AddNewRuleProps = {
  children?: PreactChildren;
  className?: string;
};

export default function AddNewRule({ children, className }: AddNewRuleProps) {
  const {
    storage: {
      sync: { matchers },
    },
  } = useConfig();
  const { showToast } = useToast();

  const handleNewRuleClick = () => {
    storageSetByKeys(
      {
        matchers: [
          ...(matchers ?? []),
          {
            active: true,
            identifier: uuidv4(),
            queries: [],
            queryPatterns: [],
            replacement: "",
            useGlobalReplacementStyle: true,
          },
        ],
      },
      {
        onError: (message) => {
          showToast({
            message,
            options: { severity: "danger" },
          });
        },
      }
    );
  };

  return (
    <button
      className={className ?? "btn btn-secondary btn-sm"}
      data-testid="add-new-rule-button"
      onClick={handleNewRuleClick}
    >
      <span className="d-flex align-items-center">
        {children ?? (
          <>
            <i className="material-icons-sharp me-1 fs-6">add</i> New rule
          </>
        )}
      </span>
    </button>
  );
}
