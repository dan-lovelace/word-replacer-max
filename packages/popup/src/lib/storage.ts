import { STORAGE_MATCHER_PREFIX } from "@worm/shared/src/browser";
import { Matcher } from "@worm/types/src/rules";
import {
  Storage,
  StorageChange,
  StorageChangeOf,
  StorageProvider,
  SyncStorage,
} from "@worm/types/src/storage";

/**
 * Converts a stored set of storage data to a shape expected by the rest of the
 * application.
 * @param currentStorage The storage object as held in the extension's storage
 * system.
 * @param changes An object containing changed keys and their new & old values.
 * @param areaName Which storage area was updated.
 * @returns For `local` and `session` storage, the keys are retained as-is and
 * any necessary changes are made before returning the new object. For `sync`,
 * any updated keys that begin with `matcher__` have their values added to a
 * list of matchers to be returned as an array.
 */
export const getUpdatedStorage = (
  currentStorage: Storage,
  changes: Record<string, StorageChange>,
  areaName: StorageProvider
) => {
  if (!areaName || !["local", "session", "sync"].includes(areaName)) {
    return currentStorage;
  }

  const area = areaName as StorageProvider;
  const prevStorageArea = { ...currentStorage[area] };
  const updatedArea = { ...prevStorageArea };

  /**
   * A helper method for changing `updatedArea` based on a given key and
   * set of changes.
   */
  const commonUpdate = (key: string, change: StorageChange) => {
    if (change.newValue === undefined) {
      // Key is being removed
      delete (updatedArea as any)[key];
    } else {
      // Key is being updated
      (updatedArea as any)[key] = change.newValue;
    }
  };

  if (area === "sync") {
    const prevSyncStorage = updatedArea as SyncStorage;

    Object.entries(changes).forEach(([storedKey, change]) => {
      if (storedKey.startsWith(STORAGE_MATCHER_PREFIX)) {
        // Updating a matcher, modify the existing array of matchers
        const { newValue, oldValue } = change as StorageChangeOf<Matcher>;

        const newMatchers: Matcher[] = [...(prevSyncStorage.matchers || [])];

        if (newValue === undefined) {
          // Rule is being deleted
          const existingIdx =
            newMatchers.findIndex(
              (matcher) => matcher.identifier === oldValue?.identifier
            ) ?? -1;

          if (existingIdx < 0) {
            throw new Error("Unable to locate existing matcher during update");
          }

          newMatchers.splice(existingIdx, 1);
        } else if (oldValue === undefined) {
          // Rule is being added
          newMatchers.push(newValue);
        } else {
          // Rule is being updated
          const existingIdx =
            newMatchers.findIndex(
              (matcher) => matcher.identifier === oldValue?.identifier
            ) ?? -1;

          if (existingIdx < 0) {
            throw new Error("Unable to locate existing matcher during update");
          }

          newMatchers[existingIdx] = newValue;
        }

        (updatedArea as SyncStorage).matchers = newMatchers;
      } else {
        commonUpdate(storedKey, change);
      }
    });
  } else {
    // All other areas
    Object.entries(changes).forEach(([storedKey, change]) => {
      commonUpdate(storedKey, change);
    });
  }

  // Return new storage object with updated area
  return {
    ...currentStorage,
    [area]: updatedArea,
  };
};
