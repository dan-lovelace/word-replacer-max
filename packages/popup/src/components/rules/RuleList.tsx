import { useConfig } from "../../store/Config";

import Alert from "../Alerts";
import RuleRow from "../rules/RuleRow";

import AddNewRule from "./AddNewRule";

export default function RuleList() {
  const {
    storage: {
      sync: { matchers },
    },
  } = useConfig();

  if (!Boolean(matchers?.length)) {
    return (
      <div className="row">
        <div className="col-12 col-lg-8 col-xxl-6">
          <Alert title="No rules">
            Your rules list is empty.{" "}
            <AddNewRule className="btn btn-link alert-link p-0 mx-1">
              Click here
            </AddNewRule>{" "}
            to create your first replacement.
          </Alert>
        </div>
      </div>
    );
  }

  return (
    <>
      <div className="row gx-3 gy-2">
        {matchers?.map((matcher) => (
          <div
            key={matcher.identifier}
            className="rule-row-wrapper col-12 col-xxl-6 position-relative"
          >
            <div className="container-fluid gx-1">
              <RuleRow matcher={matcher} matchers={matchers} />
            </div>
          </div>
        ))}
      </div>
      <div className="ps-5 pt-2">
        <AddNewRule />
      </div>
    </>
  );
}
