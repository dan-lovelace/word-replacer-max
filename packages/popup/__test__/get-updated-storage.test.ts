import { merge } from "ts-deepmerge";

import { expect } from "@jest/globals";

import { getUpdatedStorage } from "@worm/popup/src/lib/storage";
import { Matcher, MatcherInSync } from "@worm/types/src/rules";
import {
  RecentSuggestions,
  SessionStorage,
  Storage,
  StorageProvider,
} from "@worm/types/src/storage";

jest.mock("@worm/shared/src/config/values", () => ({
  VITE_API_ORIGIN: "https://dev-api.wordreplacermax.com",
}));

const TEST_MATCHER_IDENTIFIER_1 = "matcher__1234";
const TEST_MATCHER_IDENTIFIER_2 = "matcher__4567";

const defaultAppMatchers: Matcher[] = [
  {
    active: true,
    identifier: TEST_MATCHER_IDENTIFIER_1,
    queries: ["lorem"],
    queryPatterns: [],
    replacement: "sit",
  },
  {
    active: true,
    identifier: TEST_MATCHER_IDENTIFIER_2,
    queries: ["dolor"],
    queryPatterns: [],
    replacement: "ipsum",
  },
];

const TEST_MATCHER_1 = defaultAppMatchers[0];
const TEST_MATCHER_2 = defaultAppMatchers[1];

const generateAppMatchers = (overrides: Matcher[] = []) => [
  ...defaultAppMatchers,
  ...overrides,
];

const defaultStoredMatchers: MatcherInSync = {
  [TEST_MATCHER_IDENTIFIER_1]: {
    active: true,
    identifier: TEST_MATCHER_IDENTIFIER_1,
    queries: ["lorem"],
    queryPatterns: [],
    replacement: "sit",
  },
  [TEST_MATCHER_IDENTIFIER_2]: {
    active: true,
    identifier: TEST_MATCHER_IDENTIFIER_2,
    queries: ["dolor"],
    queryPatterns: [],
    replacement: "ipsum",
  },
};

const generateStoredMatchers = (overrides: MatcherInSync = {}) =>
  merge(defaultStoredMatchers, overrides);

const defaultRecentSuggestions: RecentSuggestions = {
  [TEST_MATCHER_IDENTIFIER_1]: {
    identifier: "1234",
    apiResponseData: {
      suggestions: [{ text: "lorem" }],
      tone: "neutral",
    },
    selectedTone: "neutral",
  },
};

const generateRecentSuggestions = (
  overrides: Partial<RecentSuggestions> = {}
) => merge(defaultRecentSuggestions, overrides) as RecentSuggestions;

const defaultSessionStorage: SessionStorage = {
  authAccessToken: "XXXX",
  authClockDrift: "10",
  authIdToken: "XXXX",
  authLastAuthUser: "test@wordreplacermax.com",
  authRefreshToken: "XXXX",
};

const generateSessionStorage = (overrides: Partial<SessionStorage> = {}) =>
  merge(defaultSessionStorage, overrides);

const defaultStorage: Storage = {
  local: {},
  session: {},
  sync: {},
};

const generateStorage = (overrides: Partial<Storage> = {}) =>
  merge(defaultStorage, overrides) as Storage;

describe("getUpdatedStorage", () => {
  describe("local area", () => {
    const TEST_AREA_NAME: StorageProvider = "local";

    it("returns an unmodified copy of storage when no changes exist", () => {
      const prevStorage: Storage = {
        local: {
          recentSuggestions: generateRecentSuggestions(),
        },
        session: {},
        sync: {},
      };

      const testResult = getUpdatedStorage(
        prevStorage,
        {
          lorem: {
            newValue: undefined,
            oldValue: undefined,
          },
        },
        TEST_AREA_NAME
      );

      expect(testResult).toMatchSnapshot();
    });

    it("adds a single value when no other keys exist", () => {
      const testResult = getUpdatedStorage(
        generateStorage(),
        {
          recentSuggestions: {
            newValue: defaultRecentSuggestions[TEST_MATCHER_IDENTIFIER_1],
            oldValue: undefined,
          },
        },
        TEST_AREA_NAME
      );

      expect(testResult).toMatchSnapshot();
    });

    it("adds a single value when other keys exist", () => {
      const testResult = getUpdatedStorage(
        generateStorage({
          [TEST_AREA_NAME]: {
            recentSuggestions: generateRecentSuggestions(),
          },
        }),
        {
          lorem: {
            newValue: "ipsum",
            oldValue: undefined,
          },
        },
        TEST_AREA_NAME
      );

      expect(testResult).toMatchSnapshot();
    });

    it("updates an existing value", () => {
      const oldValue = generateRecentSuggestions();

      const newValue = merge(oldValue, {
        [TEST_MATCHER_IDENTIFIER_1]: { selectedTone: "casual" },
      });

      const testResult = getUpdatedStorage(
        generateStorage({
          [TEST_AREA_NAME]: {
            recentSuggestions: oldValue,
          },
        }),
        {
          recentSuggestions: {
            newValue,
            oldValue,
          },
        },
        TEST_AREA_NAME
      );

      expect(testResult).toMatchSnapshot();
    });

    it("removes items when the new value is undefined", () => {
      const oldValue = generateRecentSuggestions();
      const testResult = getUpdatedStorage(
        generateStorage({
          [TEST_AREA_NAME]: {
            recentSuggestions: oldValue,
          },
        }),
        {
          recentSuggestions: {
            newValue: undefined,
            oldValue,
          },
        },
        TEST_AREA_NAME
      );

      expect(testResult).toMatchSnapshot();
    });
  });

  describe("session area", () => {
    const TEST_AREA_NAME: StorageProvider = "session";

    it("returns an unmodified copy of storage when no changes exist", () => {
      const prevStorage: Storage = {
        local: {},
        session: generateSessionStorage(),
        sync: {},
      };

      const testResult = getUpdatedStorage(
        prevStorage,
        {
          lorem: {
            newValue: undefined,
            oldValue: undefined,
          },
        },
        TEST_AREA_NAME
      );

      expect(testResult).toMatchSnapshot();
    });

    it("adds a single value when no other keys exist", () => {
      const prevStorage = generateStorage();

      const testResult = getUpdatedStorage(
        prevStorage,
        {
          authAccessToken: {
            newValue: "YYYY",
            oldValue: undefined,
          },
        },
        TEST_AREA_NAME
      );

      expect(testResult).toMatchSnapshot();
    });

    it("adds a single value when other keys exist", () => {
      const prevStorage: Storage = {
        local: {},
        session: generateSessionStorage(),
        sync: {},
      };

      const testResult = getUpdatedStorage(
        prevStorage,
        {
          lorem: {
            newValue: "ipsum",
            oldValue: undefined,
          },
        },
        TEST_AREA_NAME
      );

      expect(testResult).toMatchSnapshot();
    });

    it("updates an existing value", () => {
      const prevStorage: Storage = {
        local: {},
        session: generateSessionStorage(),
        sync: {},
      };

      const testResult = getUpdatedStorage(
        prevStorage,
        {
          authAccessToken: {
            newValue: "YYYY",
            oldValue: undefined,
          },
        },
        TEST_AREA_NAME
      );

      expect(testResult).toMatchSnapshot();
    });

    it("removes items when the new value is undefined", () => {
      const prevStorage: Storage = {
        local: {},
        session: generateSessionStorage(),
        sync: {},
      };

      const testResult = getUpdatedStorage(
        prevStorage,
        {
          authAccessToken: {
            newValue: undefined,
            oldValue: undefined,
          },
        },
        TEST_AREA_NAME
      );

      expect(testResult).toMatchSnapshot();
    });
  });

  describe("sync area", () => {
    const TEST_AREA_NAME: StorageProvider = "sync";

    describe("non-matcher changes", () => {
      it("returns an unmodified copy of storage when no changes exist", () => {
        const prevStorage: Storage = {
          local: {},
          session: {},
          sync: {
            storageVersion: "1.0.0",
          },
        };

        const testResult = getUpdatedStorage(
          prevStorage,
          {
            lorem: {
              newValue: undefined,
              oldValue: undefined,
            },
          },
          TEST_AREA_NAME
        );

        expect(testResult).toMatchSnapshot();
      });

      it("adds a single value when no other keys exist", () => {
        const prevStorage = generateStorage();

        const testResult = getUpdatedStorage(
          prevStorage,
          {
            storageVersion: {
              newValue: "1.1.1",
              oldValue: undefined,
            },
          },
          TEST_AREA_NAME
        );

        expect(testResult).toMatchSnapshot();
      });

      it("adds a single value when other keys exist", () => {
        const prevStorage: Storage = {
          local: {},
          session: {},
          sync: {
            storageVersion: "1.0.0",
          },
        };

        const testResult = getUpdatedStorage(
          prevStorage,
          {
            lorem: {
              newValue: "ipsum",
              oldValue: undefined,
            },
          },
          TEST_AREA_NAME
        );

        expect(testResult).toMatchSnapshot();
      });

      it("updates an existing value", () => {
        const prevStorage: Storage = {
          local: {},
          session: {},
          sync: {
            storageVersion: "1.0.0",
          },
        };

        const testResult = getUpdatedStorage(
          prevStorage,
          {
            storageVersion: {
              newValue: "1.1.1",
              oldValue: "1.0.0",
            },
          },
          TEST_AREA_NAME
        );

        expect(testResult).toMatchSnapshot();
      });

      it("removes items when the new value is undefined", () => {
        const prevStorage: Storage = {
          local: {},
          session: {},
          sync: {
            storageVersion: "1.0.0",
          },
        };

        const testResult = getUpdatedStorage(
          prevStorage,
          {
            storageVersion: {
              newValue: undefined,
              oldValue: "1.0.0",
            },
          },
          TEST_AREA_NAME
        );

        expect(testResult).toMatchSnapshot();
      });
    });

    describe("matcher changes", () => {
      it("returns an unmodified copy of storage when no changes exist", () => {
        const prevStorage: Storage = {
          local: {},
          session: {},
          sync: {
            matchers: generateAppMatchers(),
          },
        };

        const testResult = getUpdatedStorage(
          prevStorage,
          {
            lorem: {
              newValue: undefined,
              oldValue: undefined,
            },
          },
          TEST_AREA_NAME
        );

        expect(testResult).toMatchSnapshot();
      });

      it("adds a single value when no other keys exist", () => {
        const prevStorage = generateStorage();

        const testResult = getUpdatedStorage(
          prevStorage,
          {
            [TEST_MATCHER_IDENTIFIER_1]: {
              newValue: generateStoredMatchers()[TEST_MATCHER_IDENTIFIER_1],
              oldValue: undefined,
            },
          },
          TEST_AREA_NAME
        );

        expect(testResult).toMatchSnapshot();
      });

      it("adds a single value when other keys exist", () => {
        const prevStorage: Storage = {
          local: {},
          session: {},
          sync: {
            matchers: [TEST_MATCHER_1],
          },
        };

        const testResult = getUpdatedStorage(
          prevStorage,
          {
            [TEST_MATCHER_IDENTIFIER_2]: {
              newValue: TEST_MATCHER_2,
              oldValue: undefined,
            },
          },
          TEST_AREA_NAME
        );

        expect(testResult).toMatchSnapshot();
      });

      it("updates an existing value", () => {
        const prevStorage: Storage = {
          local: {},
          session: {},
          sync: {
            matchers: [TEST_MATCHER_1],
          },
        };

        const testResult = getUpdatedStorage(
          prevStorage,
          {
            [TEST_MATCHER_IDENTIFIER_1]: {
              newValue: { ...TEST_MATCHER_1, active: false },
              oldValue: TEST_MATCHER_1,
            },
          },
          TEST_AREA_NAME
        );

        expect(testResult).toMatchSnapshot();
      });

      it("removes items when the new value is undefined", () => {
        const prevStorage: Storage = {
          local: {},
          session: {},
          sync: {
            matchers: [TEST_MATCHER_1],
          },
        };

        const testResult = getUpdatedStorage(
          prevStorage,
          {
            [TEST_MATCHER_IDENTIFIER_1]: {
              newValue: undefined,
              oldValue: TEST_MATCHER_1,
            },
          },
          TEST_AREA_NAME
        );

        expect(testResult).toMatchSnapshot();
      });
    });
  });

  describe("multiple storage areas", () => {
    it("handles changes across all storage areas simultaneously", () => {
      const prevStorage: Storage = {
        local: {
          recentSuggestions: generateRecentSuggestions(),
        },
        session: generateSessionStorage(),
        sync: {
          matchers: [TEST_MATCHER_1],
        },
      };

      const oldLocalValue = defaultRecentSuggestions[TEST_MATCHER_IDENTIFIER_1];
      const withLocalChanges = getUpdatedStorage(
        prevStorage,
        {
          recentSuggestions: {
            newValue: {
              ...oldLocalValue,
              identifier: "4567",
            },
            oldValue: oldLocalValue,
          },
        },
        "local"
      );

      const finalResult = getUpdatedStorage(
        getUpdatedStorage(
          withLocalChanges,
          {
            authAccessToken: {
              newValue: "ZZZZ",
              oldValue: "XXXX",
            },
          },
          "session"
        ),
        {
          [TEST_MATCHER_IDENTIFIER_2]: {
            newValue: TEST_MATCHER_2,
            oldValue: undefined,
          },
        },
        "sync"
      );

      expect(finalResult).toMatchSnapshot();
    });

    it("handles multiple changes within each storage area", () => {
      const prevStorage: Storage = {
        local: {
          recentSuggestions: generateRecentSuggestions(),
        },
        session: generateSessionStorage(),
        sync: {
          matchers: [TEST_MATCHER_1],
          storageVersion: "1.0.0",
        },
      };

      const withLocalChanges = getUpdatedStorage(
        prevStorage,
        {
          recentSuggestions: {
            newValue: undefined,
            oldValue: generateRecentSuggestions(),
          },
          "new-local-key": {
            newValue: "new-value",
            oldValue: undefined,
          },
        },
        "local"
      );

      const withSessionChanges = getUpdatedStorage(
        withLocalChanges,
        {
          authAccessToken: {
            newValue: "ZZZZ",
            oldValue: "XXXX",
          },
          authIdToken: {
            newValue: undefined,
            oldValue: "XXXX",
          },
          newSessionKey: {
            newValue: "session-value",
            oldValue: undefined,
          },
        },
        "session"
      );

      const finalResult = getUpdatedStorage(
        withSessionChanges,
        {
          [TEST_MATCHER_IDENTIFIER_1]: {
            newValue: undefined,
            oldValue: TEST_MATCHER_1,
          },
          [TEST_MATCHER_IDENTIFIER_2]: {
            newValue: TEST_MATCHER_2,
            oldValue: undefined,
          },
          storageVersion: {
            newValue: "2.0.0",
            oldValue: "1.0.0",
          },
        },
        "sync"
      );

      expect(finalResult).toMatchSnapshot();
    });

    it("handles empty changes in some areas while updating others", () => {
      const prevStorage: Storage = {
        local: {
          recentSuggestions: generateRecentSuggestions(),
        },
        session: generateSessionStorage(),
        sync: {
          matchers: [TEST_MATCHER_1],
        },
      };

      const withLocalChanges = getUpdatedStorage(
        prevStorage,
        {
          someKey: {
            newValue: undefined,
            oldValue: undefined,
          },
        },
        "local"
      );

      const withSessionChanges = getUpdatedStorage(
        withLocalChanges,
        {
          authAccessToken: {
            newValue: "ZZZZ",
            oldValue: "XXXX",
          },
        },
        "session"
      );

      const finalResult = getUpdatedStorage(
        withSessionChanges,
        {
          anotherKey: {
            newValue: undefined,
            oldValue: undefined,
          },
        },
        "sync"
      );

      expect(finalResult).toMatchSnapshot();
    });

    it("handles complex matcher updates while modifying other storage areas", () => {
      const prevStorage: Storage = {
        local: {
          recentSuggestions: generateRecentSuggestions(),
        },
        session: generateSessionStorage(),
        sync: {
          matchers: [TEST_MATCHER_1, TEST_MATCHER_2],
        },
      };

      const oldLocalValue = defaultRecentSuggestions[TEST_MATCHER_IDENTIFIER_1];
      const withLocalChanges = getUpdatedStorage(
        prevStorage,
        {
          recentSuggestions: {
            newValue: {
              ...oldLocalValue,
              selectedTone: "professional",
            },
            oldValue: oldLocalValue,
          },
        },
        "local"
      );

      const withSessionChanges = getUpdatedStorage(
        withLocalChanges,
        {
          authLastAuthUser: {
            newValue: "test-changed@wordreplacermax.com",
            oldValue: "test@wordreplacermax.com",
          },
        },
        "session"
      );

      const updatedMatcher1 = {
        ...TEST_MATCHER_1,
        active: false,
        queries: ["updated-lorem"],
      };

      const updatedMatcher2 = {
        ...TEST_MATCHER_2,
        replacement: "updated-ipsum",
      };

      const finalResult = getUpdatedStorage(
        withSessionChanges,
        {
          [TEST_MATCHER_IDENTIFIER_1]: {
            newValue: updatedMatcher1,
            oldValue: TEST_MATCHER_1,
          },
          [TEST_MATCHER_IDENTIFIER_2]: {
            newValue: updatedMatcher2,
            oldValue: TEST_MATCHER_2,
          },
        },
        "sync"
      );

      expect(finalResult).toMatchSnapshot();
    });
  });
});
