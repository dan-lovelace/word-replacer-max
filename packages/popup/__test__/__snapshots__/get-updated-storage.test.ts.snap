// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getUpdatedStorage local area adds a single value when no other keys exist 1`] = `
{
  "local": {
    "recentSuggestions": {
      "apiResponseData": {
        "suggestions": [
          {
            "text": "lorem",
          },
        ],
        "tone": "neutral",
      },
      "identifier": "1234",
      "selectedTone": "neutral",
    },
  },
  "session": {},
  "sync": {},
}
`;

exports[`getUpdatedStorage local area adds a single value when other keys exist 1`] = `
{
  "local": {
    "lorem": "ipsum",
    "recentSuggestions": {
      "matcher__1234": {
        "apiResponseData": {
          "suggestions": [
            {
              "text": "lorem",
            },
          ],
          "tone": "neutral",
        },
        "identifier": "1234",
        "selectedTone": "neutral",
      },
    },
  },
  "session": {},
  "sync": {},
}
`;

exports[`getUpdatedStorage local area removes items when the new value is undefined 1`] = `
{
  "local": {},
  "session": {},
  "sync": {},
}
`;

exports[`getUpdatedStorage local area returns an unmodified copy of storage when no changes exist 1`] = `
{
  "local": {
    "recentSuggestions": {
      "matcher__1234": {
        "apiResponseData": {
          "suggestions": [
            {
              "text": "lorem",
            },
          ],
          "tone": "neutral",
        },
        "identifier": "1234",
        "selectedTone": "neutral",
      },
    },
  },
  "session": {},
  "sync": {},
}
`;

exports[`getUpdatedStorage local area updates an existing value 1`] = `
{
  "local": {
    "recentSuggestions": {
      "matcher__1234": {
        "apiResponseData": {
          "suggestions": [
            {
              "text": "lorem",
            },
          ],
          "tone": "neutral",
        },
        "identifier": "1234",
        "selectedTone": "casual",
      },
    },
  },
  "session": {},
  "sync": {},
}
`;

exports[`getUpdatedStorage session area adds a single value when no other keys exist 1`] = `
{
  "local": {},
  "session": {
    "authAccessToken": "YYYY",
  },
  "sync": {},
}
`;

exports[`getUpdatedStorage session area adds a single value when other keys exist 1`] = `
{
  "local": {},
  "session": {
    "authAccessToken": "XXXX",
    "authClockDrift": "10",
    "authIdToken": "XXXX",
    "authLastAuthUser": "test@wordreplacermax.com",
    "authRefreshToken": "XXXX",
    "lorem": "ipsum",
  },
  "sync": {},
}
`;

exports[`getUpdatedStorage session area removes items when the new value is undefined 1`] = `
{
  "local": {},
  "session": {
    "authClockDrift": "10",
    "authIdToken": "XXXX",
    "authLastAuthUser": "test@wordreplacermax.com",
    "authRefreshToken": "XXXX",
  },
  "sync": {},
}
`;

exports[`getUpdatedStorage session area returns an unmodified copy of storage when no changes exist 1`] = `
{
  "local": {},
  "session": {
    "authAccessToken": "XXXX",
    "authClockDrift": "10",
    "authIdToken": "XXXX",
    "authLastAuthUser": "test@wordreplacermax.com",
    "authRefreshToken": "XXXX",
  },
  "sync": {},
}
`;

exports[`getUpdatedStorage session area updates an existing value 1`] = `
{
  "local": {},
  "session": {
    "authAccessToken": "YYYY",
    "authClockDrift": "10",
    "authIdToken": "XXXX",
    "authLastAuthUser": "test@wordreplacermax.com",
    "authRefreshToken": "XXXX",
  },
  "sync": {},
}
`;

exports[`getUpdatedStorage sync area matcher changes adds a single value when no other keys exist 1`] = `
{
  "local": {},
  "session": {},
  "sync": {
    "matchers": [
      {
        "active": true,
        "identifier": "matcher__1234",
        "queries": [
          "lorem",
        ],
        "queryPatterns": [],
        "replacement": "sit",
      },
    ],
  },
}
`;

exports[`getUpdatedStorage sync area matcher changes adds a single value when other keys exist 1`] = `
{
  "local": {},
  "session": {},
  "sync": {
    "matchers": [
      {
        "active": true,
        "identifier": "matcher__1234",
        "queries": [
          "lorem",
        ],
        "queryPatterns": [],
        "replacement": "sit",
      },
      {
        "active": true,
        "identifier": "matcher__4567",
        "queries": [
          "dolor",
        ],
        "queryPatterns": [],
        "replacement": "ipsum",
      },
    ],
  },
}
`;

exports[`getUpdatedStorage sync area matcher changes removes items when the new value is undefined 1`] = `
{
  "local": {},
  "session": {},
  "sync": {
    "matchers": [],
  },
}
`;

exports[`getUpdatedStorage sync area matcher changes returns an unmodified copy of storage when no changes exist 1`] = `
{
  "local": {},
  "session": {},
  "sync": {
    "matchers": [
      {
        "active": true,
        "identifier": "matcher__1234",
        "queries": [
          "lorem",
        ],
        "queryPatterns": [],
        "replacement": "sit",
      },
      {
        "active": true,
        "identifier": "matcher__4567",
        "queries": [
          "dolor",
        ],
        "queryPatterns": [],
        "replacement": "ipsum",
      },
    ],
  },
}
`;

exports[`getUpdatedStorage sync area matcher changes updates an existing value 1`] = `
{
  "local": {},
  "session": {},
  "sync": {
    "matchers": [
      {
        "active": false,
        "identifier": "matcher__1234",
        "queries": [
          "lorem",
        ],
        "queryPatterns": [],
        "replacement": "sit",
      },
    ],
  },
}
`;

exports[`getUpdatedStorage sync area non-matcher changes adds a single value when no other keys exist 1`] = `
{
  "local": {},
  "session": {},
  "sync": {
    "storageVersion": "1.1.1",
  },
}
`;

exports[`getUpdatedStorage sync area non-matcher changes adds a single value when other keys exist 1`] = `
{
  "local": {},
  "session": {},
  "sync": {
    "lorem": "ipsum",
    "storageVersion": "1.0.0",
  },
}
`;

exports[`getUpdatedStorage sync area non-matcher changes removes items when the new value is undefined 1`] = `
{
  "local": {},
  "session": {},
  "sync": {},
}
`;

exports[`getUpdatedStorage sync area non-matcher changes returns an unmodified copy of storage when no changes exist 1`] = `
{
  "local": {},
  "session": {},
  "sync": {
    "storageVersion": "1.0.0",
  },
}
`;

exports[`getUpdatedStorage sync area non-matcher changes updates an existing value 1`] = `
{
  "local": {},
  "session": {},
  "sync": {
    "storageVersion": "1.1.1",
  },
}
`;
